### DISCUSSION TITLE
Add get-string and get-type functions

### DISCUSSION THREAD

**Discussion by Denis535** [stylus/stylus#2836] [Show and tell]
function getString(expr) {{
                expr = Stylus.utils.unwrap(expr);
                if (expr.nodes.length == 0) {{
                    return new Stylus.nodes.Null();
                }}
                if (expr.nodes.length == 1) {{
                    const value = expr.nodes[0];
                    if (value.constructor.name == 'String')   return new Stylus.nodes.String( value.string ?? value.toString() );
                    if (value.constructor.name == 'Literal')  return new Stylus.nodes.String( value.string ?? value.toString() );
                    if (value.constructor.name == 'Ident')    return new Stylus.nodes.String( value.string ?? value.toString() );
                    if (value.constructor.name == 'Unit')     return new Stylus.nodes.String( value.string ?? value.toString() );
                    if (value.constructor.name == 'RGBA')     return new Stylus.nodes.String( value.name                       );
                    if (value.constructor.name == 'Function') return new Stylus.nodes.String( value.name                       );
                    if (value.constructor.name == 'Null')     return new Stylus.nodes.Null();
                    throw new Error( ""Argument is invalid: "" + value );
                }}
                throw new Error( ""Argument is invalid: "" + expr );
            }}

            function getType(expr) {{
                expr = Stylus.utils.unwrap(expr);
                if (expr.nodes.length == 0) {{
                    return new Stylus.nodes.Null();
                }}
                if (expr.nodes.length == 1) {{
                    const value = expr.nodes[0];
                    return new Stylus.nodes.String( value.constructor.name );
                }}
                {{
                    return new Stylus.nodes.String( expr.constructor.name );
                }}
            }}

**Comment by Denis535** [comment]
Also those functions would be useful.

            function define(name, expr, global) {{
                name = Stylus.utils.unwrap(name).nodes[0].string;
                expr = Stylus.utils.unwrap(expr);
                global = global && global.toBoolean().isTrue;
                if (global === true) {{
                    const node = new Stylus.nodes.Ident(name, expr);
                    this.global.scope.add(node);
                    return;
                }}
                if (global === false || global === undefined) {{
                    const node = new Stylus.nodes.Ident(name, expr);
                    this.currentScope.add(node);
                    return;
                }}
            }}
            
            function isDefined(name, global) {{
                name = Stylus.utils.unwrap(name).nodes[0].string;
                global = global && global.toBoolean().isTrue;
                if (global === true) {{
                    return new Stylus.nodes.Boolean(this.global.scope.locals[name]);
                }}
                if (global === false) {{
                    return new Stylus.nodes.Boolean(this.currentScope.locals[name]);
                }}
                return new Stylus.nodes.Boolean(this.lookup(name));
            }}
            
            function getDefinition(name, global) {{
                name = Stylus.utils.unwrap(name).nodes[0].string;
                global = global && global.toBoolean().isTrue;
                if (global === true) {{
                    return this.global.scope.locals[name];
                }}
                if (global === false) {{
                    return this.currentScope.locals[name];
                }}
                return this.lookup(name);
            }}
            
            function getDefinitions(global) {{
                global = global && global.toBoolean().isTrue;
                if (global === true) {{
                    const result = new Stylus.nodes.Object();
                    for (const [key, value] of Object.entries(this.global.scope.locals)) {{
                        result.setValue(key, value);
                    }}
                    return result;
                }}
                if (global === false) {{
                    const result = new Stylus.nodes.Object();
                    for (const [key, value] of Object.entries(this.currentScope.locals)) {{
                        result.setValue(key, value);
                    }}
                    return result;
                }}
                console.error('Not implemented');
            }}

**Comment by slightlyfaulty** [comment]
There's already a `typeof()` function and an `is defined` binary operator.

You can also cast values to string with the `join()` function, e.g. `join('', red)` => `'#ff0'`.

**Comment by Denis535** [comment]
> `is defined` binary operator.

It doesn't work with variables. That is, it takes variable's name but not variable's value.

The ```typeof()``` also works badly.

**Reply by slightlyfaulty** [reply]
You don't define values, you define variables. It makes no sense to check if a "value" is defined. If you want to check if a variable is defined by a string name, you can just use `lookup()`.

In future it would be a good idea to give clear examples of _why_ you think your custom functions are useful. Nothing is useful without a use-case.

**Reply by Denis535** [reply]
I mean next:
```
my-variable = null
name = "my-variable"
name is defined // It checks "name" is defined or not. But I need to check  "my-variable".
```

I can not use ```lookup()``` because defined variable can also have ```null``` value.

**Reply by Denis535** [reply]
Also ```define``` function can not define variable with array value. It can define variable only with one value.

**Reply by slightlyfaulty** [reply]
If you're using dynamic variables that much you should probably re-think your code design. For example by using [hashes](https://stylus-lang.com/docs/hashes.html#hashes) which make working with string keys a lot simpler.

**Reply by Denis535** [reply]
I use dynamic variables because I can change dynamically name of variable. For example: ```theme--color----------light``` I change on ```theme--color--light```. When there are a lot of variables, this makes a lot of sense.
